src/frame.rs

```diff
@@ -0,0 +1,392 @@
+            nad,
+            READ_BY_IDENTIFIER_SID,
+            &[
+                identifier.into(),
+                (supplier_id & 0xFF) as u8,
+                (supplier_id >> 8) as u8,
+                (function_id & 0xFF) as u8,
+                (function_id >> 8) as u8,
+            ],
+        )
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    struct CheckSumTestData<'a> {
+        pid: PID,
+        data: &'a [u8],
+        checksum: u8,
+    }
+
+    #[test]
+    fn test_enhanced_checksum() {
+        let test_data = [
+            CheckSumTestData {
+                pid: PID(0xDD),
+                data: &[0x01],
+                checksum: 0x21,
+            },
+            CheckSumTestData {
+                pid: PID(0x4A),
+                data: &[0x55, 0x93, 0xE5],
+                checksum: 0xE6,
+            },
+            CheckSumTestData {
+                pid: PID(0xBF),
+                data: &[0x40, 0xFF],
+                checksum: 0x00,
+            },
+        ];
+        for d in &test_data {
+            assert_eq!(d.checksum, checksum(d.pid, d.data));
+        }
+    }
+
+    #[test]
+    fn test_classic_checksum() {
+        let test_data = [
+            CheckSumTestData {
+                pid: PID::from_id(0x3C),
+                data: &[0x01],
+                checksum: 0xFE,
+            },
+            CheckSumTestData {
+                pid: PID::from_id(0x3D),
+                data: &[0x01],
+                checksum: 0xFE,
+            },
+            CheckSumTestData {
+                pid: PID::from_id(0x3d),
+                data: &[0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08],
+                checksum: 0xDB,
+            },
+        ];
+        for d in &test_data {
+            assert_eq!(d.checksum, classic_checksum(d.data));
+        }
+    }
+
+    #[test]
+    fn test_pid_from_id() {
+        let test_data = [
+            (0, PID(0x80)),
+            (1, PID(0xC1)),
+            (2, PID(0x42)),
+            (25, PID(0x99)),
+            (27, PID(0x5B)),
+            (29, PID(0xDD)),
+        ];
+
+        for d in &test_data {
+            let pid = PID::from_id(d.0);
+            assert_eq!(pid, d.1);
+            assert_eq!(pid.get_id(), d.0);
+        }
+    }
+
+    #[test]
+    fn test_id_uses_classic_checksum() {
+        let test_ids: &[u8] = &[0, 1, 59, 60, 63];
+
+        for i in test_ids {
+            assert_eq!(PID::from_id(*i).uses_classic_checksum(), *i >= 60);
+        }
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_pid_from_id_panic() {
+        PID::from_id(64);
+    }
+
+    #[test]
+    fn test_transport_frame() {
+        const LIN_ID_SERIAL_REQ_PAYLOAD: &[u8] = &[0x10, 0x06, 0xB2, 0x01, 0xB3, 0x00, 0x01, 0x10];
+        let frame = transport::create_single_frame(
+            diagnostic::MASTER_REQUEST_FRAME_PID,
+            transport::NAD(0x10),
+            transport::SID(0xB2),
+            &[0x01, 0xB3, 0x00, 0x01, 0x10],
```

Comment:
could you add a test to check that that the `0xff` padding works (one with no data and maybe one with 0 < x.len() < 5)?

Reply:
Added more tests.